deepsubtree.go

```diff
@@ -0,0 +1,271 @@
+	hash, err := leaf._hash()
+	if err != nil {
+		return err
+	}
+
+	n := NewNode(leaf.key, leaf.value, leaf.version)
+	n._hash()
+	err = dst.ndb.SaveNode(n)
+	if err != nil {
+		return err
+	}
+
+	prevHash := n.hash
+	for i := len(pl) - 1; i >= 0; i-- {
+		pin := pl[i]
+		hash, err = pin.Hash(hash)
+		if err != nil {
+			return err
+		}
+		if pin.Left == nil {
+			pin.Left = prevHash
+		} else if pin.Right == nil {
+			pin.Right = prevHash
+		}
+		n := &Node{
+			subtreeHeight: pin.Height,
+			size:          pin.Size,
+			version:       pin.Version,
+			leftHash:      pin.Left,
+			rightHash:     pin.Right,
+			hash:          hash,
+		}
+		prevHash = hash
+
+		has, err := dst.ndb.Has(n.hash)
+		if err != nil {
+			return err
+		}
+		if !has {
+			err = dst.ndb.SaveNode(n)
+			if err != nil {
+				return err
+			}
+		}
+		if i == 0 {
+			if err := dst.ndb.Commit(); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Traverses in the nodes in the NodeDB in the Deep Subtree
+// and links them together using the populated left and right
+// hashes and sets the root to be the node with the given rootHash
+func (dst *DeepSubTree) BuildTree(rootHash []byte) error {
+	nodes, traverseErr := dst.ndb.nodes()
+	if traverseErr != nil {
+		return fmt.Errorf("could not traverse nodedb: %w", traverseErr)
+	}
+	// Traverse through nodes and link them correctly
+	for _, node := range nodes {
+		pnode, _ := dst.ndb.GetNode(node.hash)
+		if len(pnode.leftHash) > 0 && pnode.leftNode == nil {
+			pnode.leftNode, _ = dst.ndb.GetNode(pnode.leftHash)
+		}
+		if len(pnode.rightHash) > 0 && pnode.rightNode == nil {
+			pnode.rightNode, _ = dst.ndb.GetNode(pnode.rightHash)
+		}
+	}
+	// Now that nodes are linked correctly, traverse again
+	// and set their keys correctly
+	for _, node := range nodes {
+		pnode, _ := dst.ndb.GetNode(node.hash)
+		if pnode.leftNode != nil {
+			pnode.key = pnode.leftNode.getHighestKey()
+		}
+
+		if pnode.rightNode != nil {
+			pnode.key = pnode.rightNode.getLowestKey()
+		}
+	}
+	if dst.root == nil {
+		rootNode, rootErr := dst.ndb.GetNode(rootHash)
+		if rootErr != nil {
+			return fmt.Errorf("could not set root of deep subtree: %w", rootErr)
+		}
+		dst.root = rootNode
+	}
+
+	return nil
+}
+
+// Set sets a key in the working tree with the given value.
+// Assumption: Node with given key already exists and is a leaf node.
+func (dst *DeepSubTree) Set(key []byte, value []byte) (updated bool, err error) {
+	if value == nil {
+		return updated, fmt.Errorf("attempt to store nil value at key '%s'", key)
+	}
+
+	dst.root, updated, err = dst.recursiveSet(dst.root, key, value)
+	dst.root.hash = nil
+	dst.root._hash()
+	return updated, err
+}
+
+// Helper method for set to traverse and find the node with given key
+// recursively.
+func (dst *DeepSubTree) recursiveSet(node *Node, key []byte, value []byte) (
+	newSelf *Node, updated bool, err error,
+) {
+	version := dst.version + 1
+
+	if node.isLeaf() {
```

Comment:
Please add comments explaining the cases for these if conditions so there is no room for guessing :-)

Reply:
I don't believe this comment was addressed.